azerothcore/azerothcore-wotlk
# Fix(Core/Warden): Fix desync with custom payloads.

## Changes Proposed:
-  Attempt at fixing the desync by implementing a cache for interrupted checks.
-  Normal Warden checks are not sent to the client if a payload interrupts Warden.
   - Interrupted Warden check results were ignored anyway (since you can't trust a forced result) so this won't have any different impact on previous systems.
   - This only happens when using modules that calls on Warden::ForceChecks so normal behavior is retained.

## Issues Addressed:

- Closes 

## SOURCE:


## Tests Performed:

- Tested by heyitsbench
- Steps tested outlined below.


## How to Test the Changes:


1. [Clone](https://github.com/AnchyDev/BreakingNewsOverride) the Breaking News module master branch.
2. Enable the module.
3. Set `Warden.ClientCheckFailAction` to 1 or 2 to enable Kick/Banning of players when receiving a failed Warden result.
4. Login and observe the Breaking News on the character screen.
5. Create a ton of characters (this re-sends the payload each time), log into the world, run around and observe no kick/ban.

## Known Issues and TODO List:


- Some users may load the character screen before Warden is initialized (somehow?) so they will not receive the breaking news payload. This is quite rare and has no side affect. It just means they won't see the breaking news.


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
